### PR TITLE
Add resource azurerm_automation_job_schedule

### LINKED ISSUE #1955: New resouce/feature. Link a azurerm schedule with a runbook
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Hi team,
This request comes from the following thread [#1711 ](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1711)

From powershell it is possible to link runbooks with a schedule created in an automation account, as suggested @holmesb.
With terraform I can not find any resource that can do it.

To be able to close the minimum cycle of possibilities that an automation account allows, I think it is necessary to have this resource available,

- Create automation account
- Create runbooks
- Create schedule
- **Create link between runbook and schedule** 
- To future, others features

### PR DESCRIPTION
**Pull request by draggeta**
For a personal project I was looking for this resource and while there are issues requesting this, there were no PRs. This PR adds the job schedule as a resource.

Closes #1955

## Remarks
This resource only supports create, delete and import. No updates are possible. Any change recreates the resource.

I'm also not sure what to do about the requiresImport test, so if it is wrong, please let me know.

## Example
```hcl
resource "azurerm_resource_group" "example" {
  name     = "tfex-automation-account"
  location = "West Europe"
}

resource "azurerm_automation_account" "example" {
  name                = "tfex-automation-account"
  location            = "${azurerm_resource_group.example.location}"
  resource_group_name = "${azurerm_resource_group.example.name}"

  sku {
    name = "Basic"
  }
}

resource "azurerm_automation_runbook" "example" {
  name                = "Get-AzureVMTutorial"
  location            = "${azurerm_resource_group.example.location}"
  resource_group_name = "${azurerm_resource_group.example.name}"
  account_name        = "${azurerm_automation_account.example.name}"
  log_verbose         = "true"
  log_progress        = "true"
  description         = "This is an example runbook"
  runbook_type        = "PowerShellWorkflow"

  publish_content_link {
    uri = "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-automation-runbook-getvms/Runbooks/Get-AzureVMTutorial.ps1"
  }
}

resource "azurerm_automation_schedule" "example" {
  name                    = "tfex-automation-schedule"
  resource_group_name     = "${azurerm_resource_group.example.name}"
  automation_account_name = "${azurerm_automation_account.example.name}"
  frequency               = "Week"
  interval                = 1
  timezone                = "Central Europe Standard Time"
  start_time              = "2014-04-15T18:00:15+02:00"
  description             = "This is an example schedule"

  advanced_schedule {
    week_days = ["Friday"]
  }
}

resource "azurerm_automation_job_schedule" "example" {
  resource_group_name     = "${azurerm_resource_group.example.name}"
  automation_account_name = "${azurerm_automation_account.example.name}"
  schedule_name           = "${azurerm_automation_schedule.example.name}"
  runbook_name            = "${azurerm_automation_runbook.example.name}"

  parameters = {
    Connection         = "AzureRunAsConnection"
    VMCount            = 10
  }
}
```

## Exports
The following items are also exported
- id: The full ID of the resource. This is generated by Azure itself, based on the job_schedule_id.
- job_schedule_id: the UUID identifying the runbook. This is generated at creation time by the resource and doesn't need to be provided by the user.

## Acceptance tests
```powershell
PS > go test -timeout 180m -v -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc" `
>> github.com\terraform-providers\terraform-provider-azurerm\azurerm `
>> -run "^TestAccAzureRMAutomationJobSchedule" -count=1
=== RUN   TestAccAzureRMAutomationJobSchedule_basic
=== PAUSE TestAccAzureRMAutomationJobSchedule_basic
=== RUN   TestAccAzureRMAutomationJobSchedule_complete
=== PAUSE TestAccAzureRMAutomationJobSchedule_complete
=== RUN   TestAccAzureRMAutomationJobSchedule_update
=== PAUSE TestAccAzureRMAutomationJobSchedule_update
=== RUN   TestAccAzureRMAutomationJobSchedule_requiresImport
--- SKIP: TestAccAzureRMAutomationJobSchedule_requiresImport (0.00s)
    resource_arm_automation_job_schedule_test.go:91: Skipping since resources aren't required to be imported
=== CONT  TestAccAzureRMAutomationJobSchedule_basic
=== CONT  TestAccAzureRMAutomationJobSchedule_complete
=== CONT  TestAccAzureRMAutomationJobSchedule_update
--- PASS: TestAccAzureRMAutomationJobSchedule_basic (67.48s)
--- PASS: TestAccAzureRMAutomationJobSchedule_complete (67.51s)
--- PASS: TestAccAzureRMAutomationJobSchedule_update (97.84s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm       103.420s

### CODE REVIEW

File: azurerm/resource_arm_automation_job_schedule.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+		Schema: map[string]*schema.Schema{
+
+			"resource_group_name": resourceGroupNameSchema(),
+
+			"automation_account_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_automation_job_schedule.go]: Could we get some validation on this property?

File: azurerm/resource_arm_automation_job_schedule.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+			"runbook_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_automation_job_schedule.go]: Could we get some validation on this property?

File: azurerm/resource_arm_automation_job_schedule.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+			"schedule_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_automation_job_schedule.go]: Could we get some validation on this property?

File: website/docs/r/automation_job_schedule.html.markdown [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  Links an Automation Runbook and Schedule.
+---
+
+# azurerm_automation_job_schedule
+
+Links an Automation Runbook and Schedule.
+
+## Example Usage
+
+```hcl
```
Comment (katbyte) [inline website/docs/r/automation_job_schedule.html.markdown]: Could we just show the basic usage with only the resource here?

A complete example can go in the examples folders if you'd like, with a link from here:
```
A full example can be found [here](url)

File: azurerm/resource_arm_automation_job_schedule_test.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	uuid "github.com/satori/go.uuid"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMAutomationJobScheduleCreate(t *testing.T) {
```
Comment (katbyte) [inline azurerm/resource_arm_automation_job_schedule_test.go]: Could we create two more tests so we would have:
```
TestAccAzureRMAutomationJobSchedule_basic
TestAccAzureRMAutomationJobSchedule_complete
TestAccAzureRMAutomationJobSchedule_update
```

where the last one has 4 steps: `basic, complete, basic check_import_state`

File: azurerm/resource_arm_automation_job_schedule.go (hunk 1 above)
Comment (draggeta) [reply azurerm/resource_arm_automation_job_schedule.go]: Added validation by adding them to the azure helpers folder. These could also be used on the automation, runbook and schedule resources as they currently have no/out-of-date validation.

File: website/docs/r/automation_job_schedule.html.markdown (hunk 4 above)
Comment (draggeta) [reply website/docs/r/automation_job_schedule.html.markdown]: Done :)

File: azurerm/resource_arm_automation_job_schedule_test.go (hunk 5 above)
Comment (draggeta) [reply azurerm/resource_arm_automation_job_schedule_test.go]: Let me know if this is correct.

**Comment by draggeta** [comment]
I needed to use rand to generate the random string as there were issues with the datetime random string. If the tests were run separately, there would be no issues, but if you run them in parallel, they would generate the same random string and conflict.

**Comment by draggeta** [comment]
There is also an issue where the parameters for the runbook have different capitalization randomly. This happens in the API. I've tested it with my own runbooks without terraform and it happens there as well. Sometimes a parameter named "Output" in my script, displays as "output" in the API while other times a parameter named "ResourceGroup" appears as "ResourceGroup" in the API.

Is a Diff func handy here?

**Comment by katbyte** [comment]
Is that a random occurrence that happens on the API side?

If that is the case they what we should do is open a bug on the azure-sdk-for-go detailing what the issue is, link to it here. Thinks like resource group names shouldn't change case randomly.

Once the issue is opened we can mark the property as case insensitive in the schema with a comment linking to the azure-sdk-for-go issue.

File: azurerm/resource_arm_automation_job_schedule.go (hunk 1 above)
Comment (katbyte) [reply azurerm/resource_arm_automation_job_schedule.go]: thanks! definitely will add them to those resources once this has been merged

File: azurerm/resource_arm_automation_job_schedule_test.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	uuid "github.com/satori/go.uuid"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMAutomationJobSchedule_basic(t *testing.T) {
+	resourceName := "azurerm_automation_job_schedule.test"
+	ri := rand.Int()
```
Comment (katbyte) [inline azurerm/resource_arm_automation_job_schedule_test.go]: We have a function for this already: 
```suggestion
	ri := tf.AccRandTimeInt()
```

File: azurerm/resource_arm_automation_job_schedule_test.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAzureRMAutomationJobSchedule_complete(t *testing.T) {
+	resourceName := "azurerm_automation_job_schedule.test"
+	ri := rand.Int()
```
Comment (katbyte) [inline azurerm/resource_arm_automation_job_schedule_test.go]: We have a function for this already: 
```suggestion
	ri := tf.AccRandTimeInt()
```

File: azurerm/resource_arm_automation_job_schedule_test.go [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAzureRMAutomationJobSchedule_update(t *testing.T) {
+	resourceName := "azurerm_automation_job_schedule.test"
+	ri := rand.Int()
```
Comment (katbyte) [inline azurerm/resource_arm_automation_job_schedule_test.go]: We have a function for this already: 
```suggestion
	ri := tf.AccRandTimeInt()
```

**Comment by draggeta** [comment]
Hi @katbyte , I've just run the tests and they run successfully here. I don't know why they failed on your end. 
When I switch to tf.AccRandTimeInt(), the tests fail as mentioned before, due to seemingly the tests generating the same random integer:
```powershell
--- FAIL: TestAccAzureRMAutomationJobSchedule_complete (77.17s)
    testing.go:568: Step 0 error: errors during apply:

        Error: automation.JobScheduleClient#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="BadRequest" Message="A job schedule for the specified runbook and schedule already exists."

          on ~\AppData\Local\Temp\tf-test676055953\main.tf line 49:
          (source code not available)
```

It could be a Windows issue where it somehow grabs the same 4 random numbers every time. For now I'll change it so hopefully it succeeds on your end.

**Comment by draggeta** [comment]
@katbyte, regarding the api data return, it's just the parameters specified. Here is a real production example I have. These are the parameters in my runbook:
```powershell
[CmdletBinding()]
param (
    [Parameter(Mandatory = $true)]
    [string]$ResourceGroupName,

    [Parameter(Mandatory = $true)]
    [string]$StorageAccountName,

    [Parameter(Mandatory = $true)]
    [string]$ContainerName,

    [Parameter(Mandatory = $true)]
    [string]$Connection,

    [Parameter(Mandatory = $true)]
    [string]$SeriesName,

    [Parameter()]
    [int]$Runtime = 3600,

    [Parameter()]
    [string]$MediaType = "mp3",

    [Parameter()]
    [int]$KeepCount = 10,

    [Parameter(Mandatory = $true)]
    [uri]$EpisodeUri,

    [Parameter(Mandatory = $true)]
    [uri]$WebhookUri
)
```
This is my tf file:
```hcl
resource "azurerm_automation_job_schedule" "musica_e_poesia" {
  resource_group_name     = "${azurerm_resource_group.recording.name}"
  automation_account_name = "automationaccountname"
  schedule_name           = "Música e Poesia"
  runbook_name            = "Get-Radio"

  parameters = {
    Connection         = "AzureRunAsConnection"
    ContainerName      = "recordings"
    MediaType          = "mp3"
    ResourceGroupName  = "resourcegroupname"
    Runtime            = 4200
    StorageAccountName = "storageaccountname"
    WebhookUri         = "https://url1"
    EpisodeUri         = "http://url2"
    KeepCount          = 10
    SeriesName         = "Música e Poesia"
  }
}
```
This is what happens every time when I apply it if I don't change the parameter cases:
```powershell
Terraform will perform the following actions:

-/+ azurerm_automation_job_schedule.musica_e_poesia (new resource required)
      id:                            "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resourcegroupname/providers/Microsoft.Automation/automationAccounts/automationaccountname/jobSchedules/10000001-1001-1001-1001-100000000001" => <computed> (forces new resource)
      automation_account_name:       "automationaccountname" => "automationaccountname"
      job_schedule_id:               "10000001-1001-1001-1001-100000000001" => <computed>
      parameters.%:                  "10" => "10"
      parameters.Connection:         "AzureRunAsConnection" => "AzureRunAsConnection"
      parameters.ContainerName:      "recordings" => "recordings"
      parameters.EpisodeUri:         "" => "http://url2" (forces new resource)
      parameters.KeepCount:          "" => "10" (forces new resource)
      parameters.MediaType:          "mp3" => "mp3"
      parameters.ResourceGroupName:  "resourcegroupname" => "resourcegroupname"
      parameters.Runtime:            "4200" => "4200"
      parameters.SeriesName:         "" => "Música e Poesia" (forces new resource)
      parameters.StorageAccountName: "tfreuwsarecording01" => "tfreuwsarecording01"
      parameters.WebhookUri:         "https://url1" => "https://url1"
      parameters.episodeUri:         "http://url2" => "" (forces new resource)
      parameters.keepCount:          "10" => "" (forces new resource)
      parameters.seriesName:         "Música e Poesia" => "" (forces new resource)
      resource_group_name:           "resourcegroupname" => "resourcegroupname"
      runbook_name:                  "Get-Radio" => "Get-Radio"
      schedule_name:                 "Música e Poesia" => "Música e Poesia"
```

**Comment by katbyte** [comment]
oh wow... that is kinda wild, and clearly a bug imho. I have opened a bug on the [azure SDK](https://github.com/Azure/azure-sdk-for-go/issues/4780) so hopefully we can get some insight there.

Aside from that this PR is approved and read to go. Monday i will talk to some of my colleagues about how best we can handle the map keys randomly shifting (setting them all to lowercase seems like a bad idea when some retain case and some don't)

**Comment by katbyte** [comment]
image

File: azurerm/resource_arm_automation_job_schedule_test.go [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+  description  = "This is a test runbook for terraform acceptance test"
+  runbook_type = "PowerShell"
+
+  publish_content_link {
+    uri = "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-automation-runbook-getvms/Runbooks/Get-AzureVMTutorial.ps1"
+  }
+
+  content = <<EOF
+  param(
+    [string]$Output = "World"
```
Comment (draggeta) [inline azurerm/resource_arm_automation_job_schedule_test.go]: Hi @katbyte, if the test parameters are added, then they probably should be added to the runbook here as well.

Comment (katbyte) [reply azurerm/resource_arm_automation_job_schedule_test.go]: 👍 

fwiw, i'm getting the same weird casing issue, except some that change for you stay the same here 😅

**Comment by draggeta** [comment]
@katbyte I'm wondering if this cannot be merged. This thing blocking the feature is basically "as designed" and it only affects the keys, not the actual values. This is always the case, no matter if it's done via the Go SDK, PowerShell, Python or directly via the API.

However, I do understand if for consistency, you prefer to move everything to case sensitive.

**Comment by katbyte** [comment]
Hi @draggeta,

Sorry for the delay looping back i've been pulled away from PRs as of late. After letting this sit for a while (and the clear intention of the service team not to fix this) I think we can get this merged. The easiest way forward would be to force parameter keys to be lowercase and then upon read save them as lowercase. I think making them case insensitive would be possible by using a custom diffsuppress as well as validation to ensure there is only one "key" of any casing... but due to terraforms case sensitive nature i'm not sure if this would work.

I pushed a quick commit to force it lowercase but running the tests i am getting:
```
Error: Error setting the draft Automation Runbook "Output-HelloWorld" (Account "acctestAA-191004223736436003" / Resource Group "acctestRG-191004223736436003"): automation.RunbookDraftClient#ReplaceContent: Failure sending request: StatusCode=404 -- Original Error: Code="ResourceNotFound" Message="The Resource 'Microsoft.Automation/automationAccounts/acctestAA-191004223736436003/runbooks/Output-HelloWorld' under resource group 'acctestRG-191004223736436003' was not found."
```

**Comment by draggeta** [comment]
Ah, nice, don't worry about it. I'll try and fix it in the way you requested.

Thanks!

**Comment by draggeta** [comment]
@katbyte there seems to be a bug in the runbook deployment. In don't know exactly what, but it seems like it's not creating or updating the resources. This doesn't only happen in my branch, but also with v 1.35 of the AzureRM module. I'll try and look into it tomorrow.

**Comment by katbyte** [comment]
Sounds good @draggeta, thank you for the update!

**Comment by draggeta** [comment]
After some further testing it seems that only the createorupdate function is broken. Reading, listing, and deleting all work fine. I'll go and test it with the SDK directly

**Comment by draggeta** [comment]
Hi @katbyte,
after too many hours spent trying to figure out what wasn't working I figured out it was the script itself in the publish link parameter:
https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-automation-runbook-getvms/Runbooks/Get-AzureVMTutorial.ps1 is not available anymore. Is it possible to put a script somewhere in this repo?

**Comment by katbyte** [comment]
@draggeta,

You should be able to use the commit/tag it was last available in. However long term it would be ideal to add a short one of our own to this repo so in the future we can link to it once merged to master (but not required to get this merged)

**Comment by draggeta** [comment]
@katbyte Should be done now. I don't know why I didn't think of that.

**Pull request merged**